Make KintoneClient AutoCloseable and ensure cursor is deleted

kentoyoshida opened the pull request:
### Overview
This pull request makes the `KintoneClient` class `AutoCloseable` and ensures that cursors are always deleted to prevent resource leaks.

### Detailed Description
- Implemented the `AutoCloseable` interface for the `KintoneClient` class.
- Added logic in the `close` method to delete cursors.
- Restricted cursor operations to `KintoneClient` only to prevent accidental misuse.
- Improved resource management by ensuring cursors are always deleted using Java's try-with-resources.

### Related Issues
This addresses potential issues with the error: "The maximum valid cursors per domain is 10." There were cases where the cursor was not deleted, leading to the possibility of reaching this limit and causing errors.

### Impact
This change ensures better resource management and prevents errors related to cursor limits. While there should be no major breaking changes for existing code, the restriction of cursor operations to `KintoneClient` may cause minor compatibility issues. Therefore, it requires careful review to ensure all cursors are properly managed.

### Testing
- Verified that cursors are correctly deleted after operations.
- Tested with multiple cursors to ensure the limit is not reached.

### Additional Notes
Please review the changes carefully to ensure compatibility with existing implementations and proper cursor management.

kentoyoshida committed:
Make KintoneClient AutoCloseable and ensure cursor is deleted

- Implemented AutoCloseable interface for KintoneClient
- Added logic to delete cursor in the close method
- Ensured proper resource management and cleanup
- Restricted cursor operations to KintoClient only

yas-okadatech commented on src/main/java/org/embulk/input/kintone/KintoneClient.java:
```diff
@@ -117,21 +116,24 @@ public CreateCursorResponseBody createCursor(final PluginTask task, final Schema
     }
 
-    public void deleteCursor(String cursor)
+    public void deleteCursor()
```
(nits)（元からなんですが）こちら外部から呼ばれてないないので private で良いかもですね。
今回のレビュー時にも外部で呼ばれて事前にdeleteCursorされるケースがないか気になったので。

kentoyoshida replied:
@yas-okadatech 
I changed the visibility of the deleteCursor method to private to encapsulate cursor deletion logic.

yas-okadatech commented on src/main/java/org/embulk/input/kintone/KintoneClient.java:
```diff
@@ -117,21 +116,24 @@ public CreateCursorResponseBody createCursor(final PluginTask task, final Schema
-            this.logger.error(e.toString());
+        if (this.cursorId != null) {
+            try {
+                this.recordClient.deleteCursor(this.cursorId);
```
二度呼ばれてしまうかもしれないので、 `cursorId = null` をセットしておいた方が安全かもです。
同じIDに `recordClient.deleteCursor(id)` を二度読んでも大丈夫かもしれませんが。

kentoyoshida replied:
@yas-okadatech 
I added logic to reset cursorId to null after successfully deleting the cursor.

yas-okadatech reviewed (commented):
@kentoyoshida 

基本的にLGTMです。
`cursorId` のクリア入れておいた方が良いかだけ確認お願いします 🙇

kentoyoshida committed:
Refactor deleteCursor method: Change to private and reset cursorId after deletion

- Changed the visibility of the deleteCursor method to private to encapsulate cursor deletion logic.
- Added logic to reset cursorId to null after successfully deleting the cursor.
- Ensured better encapsulation and state management within the KintoneClient class.

yas-okadatech reviewed (approved):
LGTM!
Thank you

kentoyoshida committed:
Fix: Prevent multiple cursor creation in KintoneClient

- Ensure KintoneClient#createCursor throws a RuntimeException if a cursor already exists to prevent creating a new cursor without deleting the previous one.
- Defined error message as a private static final field.
- Added a test case in TestKintoneClient to verify that an exception is thrown when trying to create a cursor while one already exists.

d-hrs reviewed (changes requested):
Even if `createCursor` is invoked multiple times, deleteCursor is only invoked once.
I think this does not result in the expected behavior, so please modify it to prevent `createCursor` from being called multiple times or invoke `deleteCursor` when `createCursor` is called.

kentoyoshida commented:
> Even if `createCursor` is invoked multiple times, deleteCursor is only invoked once. I think this does not result in the expected behavior, so please modify it to prevent `createCursor` from being called multiple times or invoke `deleteCursor` when `createCursor` is called.

Thank you for your feedback.

I understand your concern regarding the behavior when `createCursor` is invoked multiple times. To address this, I have implemented a check within the `createCursor` method to ensure that a `RuntimeException` is thrown if a cursor already exists. This change ensures that a new cursor is not created without deleting the previous one.

Here is the summary of the changes made:

* Added a check in `KintoneClient#createCursor` to throw a `RuntimeException` if a cursor already exists.
* Added a test case to verify that the exception is thrown when attempting to create a cursor while one already exists.

By implementing this check, I prevent createCursor from being called multiple times without proper cleanup, ensuring that the expected behavior is maintained.

Please review the commit for details:
[[Commit URL]](https://github.com/trocco-io/embulk-input-kintone/pull/46/commits/c187825cb1f958f7facdea5456a181f676857970)

Let me know if there are any further adjustments needed or if you have additional suggestions.

d-hrs reviewed (approved):
LGTM
